### ISSUE TITLE
NativeAOT is much slower that JIT version

### ISSUE THREAD

**Issue by Neo-vortex** [dotnet/aspnetcore#42221]

**Comment by MichalStrehovsky** [comment]
Sure we can keep this open. But do you expect the fix to be in this repo?

Looking at the ObjectMethodExecutor, the fix is basically (simplified):

```
    public object? Execute(object target, object?[]? parameters)
    {
        if (LINQ is compiled)
        {
            Debug.Assert(_executor != null, "Sync execution is not supported.");
            return _executor(target, parameters);
        }
        else
        {
            // New code
            return MethodInfo.Invoke(target, parameters);
        }
    }
```

**Comment by davidfowl** [comment]
@MichalStrehovsky yes, something like that. Or possibly not even using the `ObjectMethodExecutor` in some cases and using method info directly. I think it also make sense to track add NativeAOT variations to some of our benchmarks so we can observe these differences when we make the fixes.

**Comment by steveharter** [comment]
Do we have stats on the full stack? It would seem reflection \ interpreted IL would be a factor, but not the only one for a 3x perf degradation.

FWIW during 5.0 I researched Blazor client perf around System.Text.Json and it turned out that interpreted IL in Mono was a bit faster than the native Mono runtime for reflection although that was a point-in-time measurement. Like expressions, System.Text.Json has both an IL.Emit and standard reflection approach depending on environment capabilities.

**Comment by MichalStrehovsky** [comment]
This is running the System.Linq.Expressions interpreter. NativeAOT doesn't have an IL interpreter like Mono has.

The IL interpreter that Mono has is pretty efficient compared to what the Expression interpreter is doing. The Mono interpreter doesn't interpret the IL directly, but converts it to an IR beforehand. So e.g. if you read a field with IL, the Mono interpreter bytecode will probably already have field offsets encoded in the IR instruction stream. Whereas reading a field with the Linq.Expression interpreter will always call `FieldInfo.GetValue` because we don't have any more efficient reflection primitives that the expression interpreter could use.

We could make the Linq.Expression interpreter faster if we had better reflection primitives, but it will never be as fast as compiling the expressions into IL, and jitting them to native code (or jitting them to a more efficient IR that is runtime-specific).

**Comment by jkoritzinsky** [comment]
@davidfowl I remember us chatting early in .NET 7 about some ideas on how to improve the reflection primitives to reduce the maintainability burden of using generics. Maybe we'll be able to explore that route more in .NET 8?

**Comment by davidfowl** [comment]
> @davidfowl I remember us chatting early in .NET 7 about some ideas on how to improve the reflection primitives to reduce the maintainability burden of using generics. Maybe we'll be able to explore that route more in .NET 8?

Yes I'd love to finish exploring that route and seeing where it leads. Basically dynamic call site generation APIs in the runtime that don't use ref emit.

**Comment by mkArtakMSFT** [comment]
@davidfowl this looks like a meta-issue tracking work on multiple teams, is it?
I'm moving it to .NET 8 Planning milestone and will let you handle how it should be tracked from there.

**Comment by davidfowl** [comment]
.NET 8 makes sense and we should leave it here
